### ISSUE TITLE
Update of D bits based on speculative update of A bits

### ISSUE THREAD

**Issue by kdockser** [riscv/riscv-svadu#14]
The Spec states  
```When two-stage address translation is active, updates of the D bit in G-stage PTEs may be performed as a result of speculative updates of the A bit in VS-stage PTEs. ```
However, nowhere does the spec state that the D bit in G-stage PTEs shall be set as a result of updates of the A bit in VS-stage PTEs. In fact, the the only time the specification states that the D bit is updated is when "the original memory access is a store". 

This sentence should be removed from the specification to avoid further confusion.

**Comment by ved-rivos** [comment]
The Privileged specification defines this:
Attempt to set A and/or D bits in VS-stage PTEs is a write for the G-stage PTE.
Section 9.5.1 : "For a memory access made to support VS-stage address translation (such as to read/write a VS-level page table), permissions are checked as though for a load or store, not for the original access type."
Section 9.6.3 : "A write pseudo-instruction (0x00002020 or 0x00003020) is used for the case that the machine is attempting automatically to update bits A and/or D in VS-level page tables."
The Svadu provides controls to enable the HW updating of A/D bits but  does not change the semantics of implicit accesses to VS-stage PTEs as incoming to G-stage as defined by the Privileged specification.

**Comment by kdockser** [comment]
Thanks Ved.

Section 9.5.1 says, as you show above, that "permissions are checked as though for a load or store, not for the _original access type_." (emphasis  mine)

However, the Svade spec says (emphasis mine)
"if the _original memory access_ is a store, also set pte.d to 1"

So, Svade is making it clear that _it_ is talking about the _original memory access_. Section 9.5.1 doesn't change this.

If I understand correctly, **9.6.3 Transformed Instruction or Pseudoinstruction for mtinst or htinst**, is talking about the pseudo instruction written into _mtinst_ or _htisnst_ on a trap into M-mode or HS-mode. If so, this section does not apply. Furthermore, it alludrs to the pseudo instruction as an "implicit memory access". 

If it is intended for Svade to write "D" in the G stage PTE for an ensuing write of "A in the VS-stage PTE, this needs to be explicitly stated. If you believe that the privileged specification already states this (I remain unconvinced), then this should be included as non-normative test. Otherwise, this will remain confusing for implementers.

Also, it appears to me that none of the models (SAIL, SPIKE, or QEMU) are following the interpretation that an attempt to set A and/or D bits in VS-stage PTEs is a write for the G-stage PTE.

**Comment by ved-rivos** [comment]
> If it is intended for Svade to write "D" in the G stage PTE for an ensuing write of "A in the VS-stage PTE, this needs to be explicitly stated. 

This is explicitly stated by the following paragraph:

```
           When a virtual page is accessed and the A bit is clear, the PTE is updated to set
           the A bit. When the virtual page is written and the D bit is clear, the PTE is updated
           to set the D bit. When G-stage address translation is in use and is not Bare, the 
           G-stage virtual pages may be accessed or written by implicit accesses to VS-level
           memory management data structures, such as page tables.
```
So when G-stage is active, the the  G-stage virtual pages may be accessed or written by implicit access to VS-stage page tables. And on access if G-stage PTE.A is clear then it is updated to set A bit and on write if D bit is clear it is updated to set the D bit.

> If you believe that the privileged specification already states this (I remain unconvinced), then this should be included as non-normative test. Otherwise, this will remain confusing for implementers.

The following are relevant sections  Further the Priv specification states in section 9.6.3
        "For guest-page faults, the trap instruction register is written with a special pseudoinstruction value
         if: (a) the fault is caused by an implicit memory access for VS-stage address translation, and (b) a
         nonzero value (the faulting guest physical address) is written to mtval2 or htval. "
       "A write pseudoinstruction (0x00002020 or 0x00003020) is used for the case that the machine is
        attempting automatically to update bits A and/or D in VS-level page tables. All other implicit
        memory accesses for VS-stage address translation will be reads. If a machine never automatically
        updates bits A or D in VS-level page tables (leaving this to software), the write case will never arise. "

Please specifically note - "if a machine never automatically updates A or D bits in VS-level page tables, the write case will never arise".  When the machine does automatically update A or D bits in VS-level pages - i.e. when `ADUE=1` - the write case occurs. 
The section 9.5.1 then states - "For a memory access made to support VS-stage address translation (such as to read/write a VS-level page table), permissions are checked as though for a load or store, not for the original access type." i.e. if the implicit memory access to VS-stage page table is a "write" to set the A and/or D bit in the VS-stage PTE then the access is a "store" to the G-stage  PTE.

**Comment by kdockser** [comment]
Thanks Ved.

The text that you cited explicitly talks about setting the "D" bit on a "write" to a virtual page. However, if you interpret "write" as any write to memory, as opposed to a **store**, it is inconsistent with the updates to the **privileged spec** translation process as detailed in the Svadu spec and which explicitly uses the phrase "original memory access is a store":


```
7. If pte.a = 0, or if the original memory access is a store and pte.d = 0:
a. If a store to pte would violate a PMA or PMP check, raise an access-fault exception
corresponding to the original access type.
b. Perform the following steps atomically:
3
i. Compare pte to the value of the PTE at address a + va.vpn[i] × PTESIZE.
ii. If the values match, set pte.a to 1 and, if the original memory access is a store, also set
pte.d to 1.
iii. If the comparison fails, return to step 2
```

Section 9.6.4 is talking about what happens on _guest-page-faults exceptions_, not the more general _guest physical address translation_ which typically occurs without a _guest-page-faults exception_. Furthermore, it refers to this is a **write pseudo instruction**, not a **store**. 

Section 9.5.1 talks about permissions for access "as though for a load or store:", it does not upgrade a "write" into a "store", not to mention upgrading a write to replace the phrase "original memory access is a store".

To summarize, the part of the Svadu spec that specifies translation uses the phrase "original memory access is a **store**" when stating when the D bit is written.
Is this part of the specification wrong? If so, we need to fix it.

At a minimum, we have an inconsistency that is causing confusion amongst implementers.

**Comment by ved-rivos** [comment]
Unfortunately the privileged specification does not have a process listed for G-stage. The section 5.3.2 of Privileged specification is written for the first stage - see step 1 starts with `satp` and not `hgatp`. However as the paragraph leading to this states "and as modified by the hypervisor extension". The H extension modifies it as follow (discussed in section 9.5.1).
" The conversion of an Sv32x4, Sv39x4, Sv48x4, or Sv57x4 guest physical address is accomplished
  with the same algorithm used for Sv32, Sv39, Sv48, or Sv57, as presented in Section 5.3.2, except
  that:
  1. hgatp substitutes for the usual satp;
  2. for the translation to begin, the effective privilege mode must be VS-mode or VU-mode;
  3. when checking the U bit, the current privilege mode is always taken to be U-mode; and
  4. guest-page-fault exceptions are raised instead of regular pag
  For a memory access made to support VS-stage address translation (such as to read/write a VS-level page table), permissions 
  are checked as though for a load or store, not for the original access type."

Following the last statement on how H-extension modifies this process - for the G-stage - the implicit memory access to read/write VS-level page table is checked as though for "for a load or store, not original access type".

**Comment by kdockser** [comment]
Thanks Ved.

Unfortunately that doesn't clearly explain the inconsistency between the more general rules in Svadu and the updates to the translation process given later in the same specification.

What I am seeing in the SAIL model is following the step-based translation process. However, even with the noted changes for G-stage translation, this doesn't match the more general rules of updating the PTE for changes made in the pointed-to page.

For the purposes of clarity, we need to expand the **Virtual Translation Process** step-based description to show the different behavior for setting the A/D bits in the G-stage PTEs. This can be done in the Svadu spec and then merged into the privileged spec. Specifically, when performing a 2-stage translation:

1. A G-stage PTE that is pointing to a VS-stage PTE that is _not_ the leaf sets its "A" bit if it is not already 1. The "D" bit is not set, even if _the original memory access was a store_.
2. A G-stage PTE that is pointing to a leaf VS-stage PTEs sets its "A" bit if it is not already 1. If the leaf VS-stage PTE is modified such that its "A" or "D" is written, then the G-stage PTE sets its "D" bit if it is not already 1.
3. A leaf G-stage PTE that provides the translation of a leaf VS-stage PTE sets its "A" bit if it is not already 1 and sets its "D" bit if if it is not already 1  and _the original memory access was a store_.

No doubt that this could be written more elegantly, but this is what needs to be said to make it clear how the 2-stage translation needs to be performed with regard to the "A" and "D" bits.

**Comment by ved-rivos** [comment]
The virtual address translation process - section 5.3.2 - is for single-stage translation. It is modified by the H-extension as noted in section 9.5.1. So for G-stage one needs to make  a copy of section 5.3.2 and then apply the modifications as provided in section 9.5.1. All these 3 points are clearly articulated in section 9.5.1. Adding them again to Priv spec do not seem required. Expanding section 5.3.2 with G-stage details is not right  since section 5.3.2 is base priv spec - the H extension modifies the base spec.

Hardware updating of the A/D bits is already part of the Priv spec as you can see in step 7 of the section 5.3.2. The step 7 in the priv spec. notes that if hardware does not cause page fault due to missing A and/or D bits it performs the steps outlined in sub bullet 2 to set the A and/or D bits atomically.

Svadu does not change when and how A/D bits are updated - this is already part of the Priv spec. Svadu only introduces a control to enable or disable this behavior.

**Comment by kdockser** [comment]
We seem to be going in circles with no further explanation added. Please answer 

How does applying the modifications of 9.5.1 to 5.3.2 change, for example,  
`ii. If the values match, set pte.a to 1 and, if the original memory access is a store, also set pte.d to 1.`?

In general, how would someone arrive at the three behaviors of the G-stage PTE that I specified above by applying these changes?

**Comment by ved-rivos** [comment]
The Guest-physical address translation modifies process in 5.3.2 as defined in section 9.5.1. The modifications are as follows:

1. hgatp substitutes for the usual satp
2. for the translation to begin, the effective privilege mode must be VS-mode or VU-mode
3. When checking the U bit, the current privilege mode is always taken to be U-mode
4. For a memory access made to support VS-stage address translation (such as to read/write a VS-level page table), permissions are checked as though for a load or store, not for the original access type However, any exception is always reported for the original access type.

- memory access is a load if G-stage invoked for read of a VS-stage page table.
- memory access is a store if G-stage invoked for write of a VS-stage page table.
- memory access is original access type if G-stage not invoked for VS-stage page table address translation.

5. The vsstatus field MXR, which makes execute-only pages readable, only overrides VS-stage page protection. Setting MXR at VS-level does not override guest-physical page protections. Setting MXR at HS-level, however, overrides both VS-stage and G-stage execute-only permissions
6. guest-page-fault exceptions are raised instead of regular page-fault exceptions

I marked up the process in 5.3.2 with locations where the modifications in 9.5.1 apply below
 
---

1. Let a be hgatp.ppn × PAGESIZE, and let i = LEVELS − 1. (For Sv32x4, PAGESIZE=2^12 and LEVELS=2.) The hgatp register must be active, i.e.,  the effective privilege mode must be S-mode or U-mode.
// Modification 1 and 2 applies to step 1
2. Let pte be the value of the PTE at address a+va.vpn[i]×PTESIZE. (For Sv32x4, PTESIZE=4.). If accessing pte violates a PMA or PMP check, raise an access-fault exception corresponding to the original access type.
3. If pte.v = 0, or if pte.r = 0 and pte.w = 1, or if any bits or encodings that are reserved for future standard use are set within pte, stop and raise a page-fault exception corresponding to the original access type.
// Modification 6 applies to exception,
4. Otherwise, the PTE is valid. If pte.r = 1 or pte.x = 1, go to step 5. Otherwise, this PTE is a pointer to the next level of the page table. Let i = i − 1. If i < 0, stop and raise a guest-page-fault exception corresponding to the original access type. Otherwise, let a = pte.ppn × PAGESIZE and go to step 2.
5. A leaf PTE has been found. Determine if the requested memory access is allowed by the pte.r, pte.w, pte.x, and pte.u bits, given the privilege mode for memory access and the value of the SUM and effective MXR. If not, stop and raise a page-fault exception corresponding to the original access type.
// Modification 3 and 5 apply to checking of pte.u, pte.r & MXR, and SUM is not applicable due to 3 and memory access type is determined by 4 and affects checking for pte.w. Modification 6 applies to exception.
6. If i > 0 and pte.ppn[i − 1 : 0] != 0, this is a misaligned superpage; stop and raise a page-fault exception corresponding to the original access type.
// modification 6 applies to the exception
7. If pte.a = 0, or if the original memory access is a store and pte.d = 0, either raise a page-fault exception corresponding to the original access type if menvcfg.HADE is 0 else: 
// The 4 modifies the memory access type - similar to how it modifies permission checking in step 5. Modification 6 applies to the exception.
- If a store to pte would violate a PMA or PMP check, raise an access-fault exception corresponding to the original access type.
- Perform the following steps atomically a) Compare pte to the value of the PTE at address a + va.vpn[i] × PTESIZE. b) If the values match, set pte.a to 1 and, if the original memory access is a store, also set pte.d to 1. c) If the comparison fails, return to step 2
// The 4 modifies the memory access type similar to how it modifies permission checking in step 5. 
9. The translation is successful. The translated physical address is given as follows:

- pa.pgoff = va.pgoff.
- If i > 0, then this is a superpage translation and pa.ppn[i − 1 : 0] = va.vpn[i − 1 : 0].
- pa.ppn[LEVELS − 1 : i] = pte.ppn[LEVELS − 1 : i]

**Comment by ved-rivos** [comment]
The process in 5.3.2 with G-stage nested as follows:
Step 1
Step 2
`   `_Add - if G-stage active, a = Translate GPA `a` for **read** of a VS-stage page table_
`   `Rest of step 2
Step 3, 4, 5, 6

Step 7 (for HADE=0)
`   `_Step 7.a - Add - if G-stage active, `a` = Translate GPA `a` for **write** of a VS-stage page table_
`   `Rest of Step 7.a - PMA/PMP check to check write permission to `a + va.vpn[i] × PTESIZE`
`   `Step 7.b
Step 8
`   `_if G-stage active, `pa` = Translate GPA `pa` for original access type_

**Comment by kdockser** [comment]
Thanks Ved.

Your addition to stage 7 of
`// The 4 modifies the memory access type similar to how it modifies permission checking in step 5`
is along the right lines.
However, 7 still says: if the _original memory access_ is a store, also set pte.d to 1.

We need to modify this to: if the _effective memory access_ is a store, also set pte.d to 1.

Then, we need to define _effective memory access_ (perhaps in 4):
- the _effective memory access_ is a LOAD if a G-stage translation points to a non-leaf VS-stage PTE
- the _effective memory access_ is a LOAD if a G-stage translation points to a leaf VS-stage PTE that is not modified by the page table walk
- the _effective memory access_ is a STORE if a G-stage translation points to a leaf VS-stage PTE whose A or D bit is set by the page table walk
- the _effective memory access_ is the original memory access if a G-stage translation points to something other than a VS-stage PTE

**Comment by ved-rivos** [comment]
> However, 7 still says: if the original memory access is a store, also set pte.d to 1.

Yes, 7 which is a step in section 5.3.2 will still say that since it is base Priv spec. The H-extension section 9.5.1 effectively writes a new process by stating that one should start from the text in 5.3.2 and modify it as per the rules in section 9.5.1. So modifying section 5.3.2 which is the Base priv spec is not right. 

> the effective memory access is a LOAD if a G-stage translation points to a non-leaf VS-stage PTE

Its  not because it points to a VS-stage PTE. It is because it the process was invoked for VS-stage address translation. There is a difference there since if guest software directly modifying a VS-stage PTE then that is not considered a "implicit access".

The "effective memory access" rules are already stated in section 9.5.1 
"Access type permissions—readable, writable, or executable—are checked during G-stage translation the same as for VS-stage translation. For a memory access made to support VS-stage address translation (such as to read/write a VS-level page table), permissions are checked as though for a load or store, not for the original access type."

Such an addition to Base Priv spec is out of scope of Svadu. I suggest an issue filed in Priv ISA manual if the explanation in section 9.5.1 is considered insufficient.

**Comment by kdockser** [comment]
The Svadu spec has its own step 7 and 8 which are meant to functionally replace step 7 in the privileged spec section 5.3.2 (as amended by 9.5.1). I had assumed that these would actually go into the Privileged Spec as a part of 1.3. In any case, it appears that step 7 in 5.3.2 is a superset of what is stated in Svadu. The only difference is the Privileged Spec allows for the option of taking a page fault exception.

Also, since "Svadu does not change when and how A/D bits are updated - this is already part of the Priv spec. Svadu only introduces a control to enable or disable this behavior." it seems that much of the Svadu spec is redundant. Will the Svadu spec be reduced to just the addition of menvcfg.HADE and henvcfg.HADE?

I had started with the assumption that Svade was going to change the Privileged Specification, including step 7. I now understand that this is not the case. I will bring these issues up for clarification with the Privileged ISA.

Thanks,
Ken

**Comment by ved-rivos** [comment]
> Will the Svadu spec be reduced to just the addition of menvcfg.HADE and henvcfg.HADE?

And the rules around H/W updating up to the text that states "Svadu extension requires the page tables to be located in cacheable main memory PMA regions."

**Comment by ved-rivos** [comment]
>  I will bring these issues up for clarification with the Privileged ISA.

Thanks.
